File: packages/sdk-python/notification_platform/_http.py
"""Shared HTTP request handling: timeouts, retries with backoff and jitter, and response
envelope parsing. Used by both NotificationClient (secret API key) and InboxClient
(end-user token) so retry/timeout behavior stays identical between the two credential types.
"""

from __future__ import annotations

import json
import random
import time
from email.utils import parsedate_to_datetime
from typing import Any, Dict, Optional
from urllib.error import HTTPError, URLError
from urllib.parse import urlencode
from urllib.request import Request, urlopen

from .errors import NotificationApiError

# GET/PUT/DELETE/PATCH are treated as safe to retry: PATCH endpoints in this API only ever
# perform partial field updates, never creates. POST is the one verb that can create a
# duplicate side effect on retry, so it is only retried when the caller supplied an
# Idempotency-Key the server can use to de-duplicate.
ALWAYS_IDEMPOTENT_METHODS = frozenset({"GET", "HEAD", "PUT", "DELETE", "PATCH"})


def _is_retryable_status(status: int) -> bool:
    return status == 429 or status >= 500


def _is_idempotent(method: str, idempotency_key: Optional[str], force_idempotent: Optional[bool]) -> bool:
    if force_idempotent is not None:
        return force_idempotent
    if method.upper() in ALWAYS_IDEMPOTENT_METHODS:
        return True
    return bool(idempotency_key)


def _backoff_delay_seconds(retry_index: int, base_seconds: float = 0.3, cap_seconds: float = 4.0) -> float:
    """Full-jitter exponential backoff: uniform(0, min(cap, base * 2^retry_index))."""
    bound = min(cap_seconds, base_seconds * (2 ** retry_index))
    return random.uniform(0, bound)


def _retry_after_seconds(header_value: Optional[str]) -> Optional[float]:
    if not header_value:
        return None
    try:
        return float(header_value)
    except ValueError:
        pass
    try:
        parsed = parsedate_to_datetime(header_value)
    except (TypeError, ValueError):
        return None
    return max(0.0, parsed.timestamp() - time.time())


def _read_json(response: Any) -> Any:
    raw = response.read()
    if not raw:
        return {}
    try:
        return json.loads(raw)
    except ValueError:
        return {}


def _extract_data(payload: Dict[str, Any]) -> Any:
    if payload.get("data") is not None:
        return payload["data"]
    # POST /v1/events replies with event_id/status/notification_ids spliced at the top level
    # instead of under `data`.
    return {key: value for key, value in payload.items() if key not in ("success", "request_id", "error")}


def _build_api_error(payload: Any, status: int) -> NotificationApiError:
    error = payload.get("error", {}) if isinstance(payload, dict) else {}
    code = error.get("code", "REQUEST_FAILED")
    message = error.get("message", f"Request failed with status {status}.")
    details = error.get("details")
    return NotificationApiError(code, message, status, details)


class HttpClient:
    """Minimal, dependency-free HTTP client built on urllib."""

    def __init__(
        self,
        base_url: str,
        authorization_header: str,
        timeout_seconds: float = 15.0,
        max_retries: int = 2,
    ) -> None:
        self.base_url = base_url.rstrip("/")
        self.authorization_header = authorization_header
        self.timeout_seconds = timeout_seconds
        self.max_retries = max_retries

    def request(
        self,
        method: str,
        path: str,
        body: Optional[Dict[str, Any]] = None,
        query: Optional[Dict[str, Any]] = None,
        idempotency_key: Optional[str] = None,
        idempotent: Optional[bool] = None,
    ) -> Any:
        url = self._build_url(path, query)
        retryable = _is_idempotent(method, idempotency_key, idempotent)
        max_attempts = 1 + (self.max_retries if retryable else 0)
        retry_after_override: Optional[float] = None
        last_network_error: Optional[Exception] = None

        for attempt in range(max_attempts):
            if attempt > 0:
                delay = retry_after_override if retry_after_override is not None else _backoff_delay_seconds(attempt - 1)
                time.sleep(delay)
                retry_after_override = None

            is_last_attempt = attempt == max_attempts - 1
            headers: Dict[str, str] = {"Authorization": self.authorization_header}
            data: Optional[bytes] = None
            if body is not None:
                headers["Content-Type"] = "application/json"
                data = json.dumps(body).encode("utf-8")
            if idempotency_key:
                headers["Idempotency-Key"] = idempotency_key

            request = Request(url, data=data, headers=headers, method=method)
            try:
                with urlopen(request, timeout=self.timeout_seconds) as response:
                    status = response.status
                    payload = _read_json(response)
            except HTTPError as error:
                status = error.code
                payload = _read_json(error)
                if not is_last_attempt and _is_retryable_status(status):
                    retry_after_override = _retry_after_seconds(error.headers.get("Retry-After"))
                    continue
                raise _build_api_error(payload, status) from error
            except (URLError, TimeoutError, OSError) as error:
                last_network_error = error
                if not is_last_attempt:
                    continue
                raise NotificationApiError("NETWORK_ERROR", str(error), 0) from error

            if status == 204 or not payload:
                return None
            if isinstance(payload, dict) and payload.get("success"):
                return _extract_data(payload)
            raise _build_api_error(payload, status)

        # Unreachable: the loop above always returns or raises before falling through.
        if last_network_error is not None:
            raise NotificationApiError("NETWORK_ERROR", str(last_network_error), 0)
        raise NotificationApiError("REQUEST_FAILED", "Request failed.", 0)

    def _build_url(self, path: str, query: Optional[Dict[str, Any]]) -> str:
        url = f"{self.base_url}{path}"
        if query:
            filtered = {key: value for key, value in query.items() if value is not None}
            if filtered:
                url = f"{url}?{urlencode(filtered)}"
        return url

File: packages/sdk-js/src/http.ts
import { NotificationApiError } from './errors';

export interface HttpRequestInit {
  method?: string;
  body?: unknown;
  query?: Record<string, string | number | boolean | undefined>;
  idempotencyKey?: string;
  /** Caller-supplied cancellation signal. Combined with (never replaced by) our own timeout signal. */
  signal?: AbortSignal;
  /**
   * Overrides the method-based idempotency guess below. Use this for endpoints that are POST
   * on the wire but semantically idempotent (e.g. the inbox "mark read" actions), so they are
   * still safe to retry even without an Idempotency-Key header.
   */
  idempotent?: boolean;
}

export interface HttpClientConfig {
  baseUrl: string;
  fetchImpl: typeof fetch;
  /** Per-request timeout in ms, enforced via AbortSignal. */
  timeoutMs: number;
  /** Number of retries *in addition to* the initial attempt. */
  maxRetries: number;
  authorizationHeader: string;
}

interface ApiEnvelope {
  success: boolean;
  data?: unknown;
  error?: { code?: string; message?: string; details?: Record<string, unknown> };
  request_id?: string;
  // POST /v1/events replies with these fields spliced at the top level instead of under `data`.
  [key: string]: unknown;
}

function isEnvelope(value: unknown): value is ApiEnvelope {
  return Boolean(value) && typeof value === 'object' && 'success' in (value as Record<string, unknown>);
}

async function safeParseJson(response: Response): Promise<unknown> {
  const text = await response.text();
  if (!text) return {};
  try {
    return JSON.parse(text);
  } catch {
    return {};
  }
}

function extractData<T>(payload: ApiEnvelope): T {
  if (payload.data !== undefined) return payload.data as T;
  const { success: _success, request_id: _requestId, error: _error, ...rest } = payload;
  return rest as T;
}

function buildApiError(payload: unknown, status: number): NotificationApiError {
  const envelope = isEnvelope(payload) ? payload : ({ success: false } as ApiEnvelope);
  return new NotificationApiError(
    envelope.error?.code ?? 'REQUEST_FAILED',
    envelope.error?.message ?? `Request failed with status ${status}.`,
    status,
    envelope.error?.details,
  );
}

function toNetworkError(error: unknown): NotificationApiError {
  const message = error instanceof Error ? error.message : 'Network request failed.';
  return new NotificationApiError('NETWORK_ERROR', message, 0);
}

function isRetryableStatus(status: number): boolean {
  return status === 429 || status >= 500;
}

// GET/PUT/DELETE are idempotent by construction; PATCH here is always a partial field update
// (never a create), so it is safe to repeat. POST is the one verb that can create a duplicate
// side effect on retry, so it is only treated as safe when the caller attached an Idempotency-Key
// the server can use to de-duplicate.
function isIdempotentRequest(method: string, idempotencyKey?: string): boolean {
  const upper = method.toUpperCase();
  if (upper === 'GET' || upper === 'HEAD' || upper === 'PUT' || upper === 'DELETE' || upper === 'PATCH') return true;
  return Boolean(idempotencyKey);
}

function retryAfterMs(header: string | null): number | undefined {
  if (!header) return undefined;
  const seconds = Number(header);
  if (!Number.isNaN(seconds)) return seconds * 1000;
  const asDate = Date.parse(header);
  if (!Number.isNaN(asDate)) return Math.max(0, asDate - Date.now());
  return undefined;
}

// Full-jitter exponential backoff: uniform(0, min(cap, base * 2^retryIndex)).
function backoffDelayMs(retryIndex: number, baseMs = 300, capMs = 4_000): number {
  const bound = Math.min(capMs, baseMs * 2 ** retryIndex);
  return Math.random() * bound;
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, Math.max(0, ms)));
}

// Node 18 (our minimum supported engine) predates AbortSignal.any (Node 20.3+), so fall back to
// a manual combinator when it is unavailable. Never used to drop the caller's own signal.
function combineSignals(primary: AbortSignal, secondary?: AbortSignal): AbortSignal {
  if (!secondary) return primary;
  if (typeof AbortSignal.any === 'function') return AbortSignal.any([primary, secondary]);
  if (primary.aborted || secondary.aborted) {
    const controller = new AbortController();
    controller.abort((primary.aborted ? primary : secondary).reason);
    return controller.signal;
  }
  const controller = new AbortController();
  const onAbort = (signal: AbortSignal) => () => controller.abort(signal.reason);
  primary.addEventListener('abort', onAbort(primary), { once: true });
  secondary.addEventListener('abort', onAbort(secondary), { once: true });
  return controller.signal;
}

function buildUrl(baseUrl: string, path: string, query?: HttpRequestInit['query']): string {
  let url = `${baseUrl}${path}`;
  if (query) {
    const params = new URLSearchParams();
    for (const [key, value] of Object.entries(query)) if (value !== undefined) params.set(key, String(value));
    const qs = params.toString();
    if (qs) url += `?${qs}`;
  }
  return url;
}

/**
 * Shared request core used by both the secret-key `NotificationClient` and the end-user
 * `InboxClient`. Neither knows about the other's credential type — this class just carries
 * whatever pre-built `authorization` header value it was configured with.
 */
export class HttpClient {
  constructor(private readonly config: HttpClientConfig) {}

  async request<T>(path: string, init: HttpRequestInit = {}): Promise<T> {
    const method = init.method ?? 'GET';
    const url = buildUrl(this.config.baseUrl, path, init.query);
    const retryable = init.idempotent ?? isIdempotentRequest(method, init.idempotencyKey);
    const maxAttempts = 1 + (retryable ? this.config.maxRetries : 0);
    let retryAfterOverrideMs: number | undefined;

    for (let attempt = 0; attempt < maxAttempts; attempt += 1) {
      if (attempt > 0) {
        await sleep(retryAfterOverrideMs ?? backoffDelayMs(attempt - 1));
        retryAfterOverrideMs = undefined;
      }

      const isLastAttempt = attempt === maxAttempts - 1;
      const timeoutController = new AbortController();
      const timer = setTimeout(() => timeoutController.abort(new DOMException('Request timed out.', 'TimeoutError')), this.config.timeoutMs);

      try {
        const response = await this.config.fetchImpl(url, {
          method,
          headers: {
            authorization: this.config.authorizationHeader,
            ...(init.body !== undefined ? { 'content-type': 'application/json' } : {}),
            ...(init.idempotencyKey ? { 'idempotency-key': init.idempotencyKey } : {}),
          },
          ...(init.body !== undefined ? { body: JSON.stringify(init.body) } : {}),
          signal: combineSignals(timeoutController.signal, init.signal),
        });
        clearTimeout(timer);

        if (response.status === 204) return undefined as T;
        const payload = await safeParseJson(response);

        if (response.ok && isEnvelope(payload) && payload.success) return extractData<T>(payload);

        const apiError = buildApiError(payload, response.status);
        if (!isLastAttempt && isRetryableStatus(response.status)) {
          retryAfterOverrideMs = retryAfterMs(response.headers.get('retry-after'));
          continue;
        }
        throw apiError;
      } catch (error) {
        clearTimeout(timer);
        if (error instanceof NotificationApiError) throw error;
        // Caller cancelled explicitly: propagate immediately, never retry over their own signal.
        if (init.signal?.aborted) throw error;
        if (!isLastAttempt) continue;
        throw toNetworkError(error);
      }
    }
    /* istanbul ignore next -- loop above always returns or throws */
    throw new NotificationApiError('REQUEST_FAILED', 'Request failed.', 0);
  }
}
